lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -1949,81 +1954,130 @@ public void realtime_presence_get_throws_when_channel_failed() throws AblyExcept
+								wrongPresenceEmitted[0] = true;
+						}
+					});
+				}
 
-			ably.connection.connectionManager.requestState(ConnectionState.suspended);
-			channelWaiter.waitFor(ChannelState.suspended);
+				final ArrayList<PresenceMessage> leaveMessages = new ArrayList<>();
+				channel.presence.subscribe(Action.leave, new Presence.PresenceListener() {
+					@Override
+					public void onPresenceMessage(PresenceMessage messages) {
+						leaveMessages.add(messages);
+					}
+				});
+
+				/*
+				 * We put testClientId2 presence data into the client library presence map but we
+				 * don't send it to the server
+				 */
+
+				MockWebsocketFactory.blockSend();
+				channel.presence.enterClient(testClientId2, presenceData);
+
+				ProtocolMessage msg = new ProtocolMessage();
+				msg.connectionId = connId;
+				msg.action = ProtocolMessage.Action.sync;
+				msg.channel = channelName;
+				msg.presence = new PresenceMessage[]{
+						new PresenceMessage() {{
+							action = Action.present;
+							id = String.format("%s:0:0", connId);
+							timestamp = System.currentTimeMillis();
+							clientId = testClientId2;
+							connectionId = connId;
+							data = presenceData;
+						}}
+				};
+				ably.connection.connectionManager.onMessage(null, msg);
+
+				MockWebsocketFactory.allowSend();
+
+				ably.connection.connectionManager.requestState(ConnectionState.suspended);
+				channelWaiter.waitFor(ChannelState.suspended);
+
+				/*
+				 * When restoring from suspended state server will send sync message erasing
+				 * testClientId2 record from the presence map. Client should re-send presence message
+				 * for testClientId2 and restore its presence data.
+				 */
+
+				ably.connection.connectionManager.requestState(ConnectionState.connected);
+				channelWaiter.waitFor(ChannelState.attached);
+				long reconnectTimestamp = System.currentTimeMillis();
+
+				try {
+					Thread.sleep(500);
+				} catch (InterruptedException e) {
+				}
 
-			/*
-			 * When restoring from suspended state server will send sync message erasing
-			 * testClientId2 record from the presence map. Client should re-send presence message
-			 * for testClientId2 and restore its presence data.
-			 */
+				AblyRest ablyRest = new AblyRest(opts);
+				io.ably.lib.rest.Channel restChannel = ablyRest.channels.get(channelName);
+				assertEquals("Verify presence data is received by the server",
+						restChannel.presence.get(null).items().length, i==0 ? 1 : 2);
 
-			ably.connection.connectionManager.requestState(ConnectionState.connected);
-			channelWaiter.waitFor(ChannelState.attached);
+				/* In both cases we should have one leave message in the leaveMessages */
+				assertEquals("Verify exactly one LEAVE message was generated", leaveMessages.size(), 1);
+				PresenceMessage leaveMessage = leaveMessages.get(0);
+				assertTrue("Verify LEAVE message follows specs",
+						leaveMessage.clientId.equals(testClientId2) && leaveMessage.id == null &&
```

Comment:
> Action type is checked implicitly on subscribe

That's not true.  That's the event that is emitted, and has nothing to do with the `PresenceMessage`.  It may do in your implementation, but that's not relevant to the tests as the library implementation can change in the future and decouple those things.